internal/controller/core/pod_controller.go

```diff
@@ -0,0 +1,196 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package core
+
+import (
+	"context"
+	"os"
+
+	"github.com/go-logr/logr"
+	"github.com/opendatahub-io/odh-model-controller/internal/controller/constants"
+	"github.com/opendatahub-io/odh-model-controller/internal/controller/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type PodReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PodReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	// Create a builder that only watch OpenDataHub global certificate ConfigMap
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Pod{}).
+		Named("core-pod").
+		WithEventFilter(reconcileServingPod()).
+		Complete(r)
+}
+
+// reconcileOpenDataHubGlobalCertConfigMap filters out all ConfigMaps that are not the OpenDataHub global certificate ConfigMap.
+func reconcileServingPod() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			pod := e.Object.(*corev1.Pod)
+			return checkPodHasIP(nil, pod)
+		},
+
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			old := e.ObjectOld.(*corev1.Pod)
+			new := e.ObjectNew.(*corev1.Pod)
+			return checkPodHasIP(old, new)
+		},
+
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+func checkPodHasIP(oldPod *corev1.Pod, newPod *corev1.Pod) bool {
+	if oldPod != nil {
+		return !(oldPod.Status.PodIP == newPod.Status.PodIP)
+	}
+
+	return newPod.Status.PodIP != ""
+}
+
+func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	// Initialize logger format
+	logger := log.FromContext(ctx).WithValues("Pod", req.Name, "namespace", req.Namespace)
+
+	controllerNs := os.Getenv("POD_NAMESPACE")
+
+	pod := &corev1.Pod{}
+	if err := r.Client.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		logger.Error(err, "Failed to get Pod", "pod", req.Name, "namespace", req.Namespace)
+		return reconcile.Result{}, err
+	}
+
+	err := r.reconcileRayTls(ctx, logger, controllerNs, req.Namespace, pod)
```

Comment:
I'm only seeing a filter at manager setup for label `component=predictor`. Here, I'm only seeing a simple check on pod IP to determine if the reconciliation should proceed.

My guess is that if the namespace holds single-node models, reconciliation will be erring if there are no multi-node models because there will be no cert secrets. If the namespace holds both single-node and multi-node models, this would generate certs for all pods, including the ones belonging to single-node. 

Would this hold true? And if so, shouldn't certs generation be limited to multi-node models pods?

Reply:
good catch! I added a check logic about it.
```
func checkMultiNodePod(pod *corev1.Pod) bool {
	for _, contatiner := range pod.Spec.Containers {
		for _, env := range contatiner.Env {
			return env.Name == "RAY_TLS_USE"
		}
	}

	return false
}
```